lib/manager/terraform/lockfile/hash.ts

```diff
@@ -0,0 +1,188 @@
+import {
+  defaultRegistryUrls,
+  id,
+} from '../../../datasource/terraform-provider';
+import type {
+  TerraformBuild,
+  VersionDetailResponse,
+} from '../../../datasource/terraform-provider/types';
+import { logger } from '../../../logger';
+import * as packageCache from '../../../util/cache/package';
+import { Http } from '../../../util/http';
+
+const http = new Http(id);
+
+export function hashFiles(files: string[]): string {
+  const rootHash = crypto.createHash('sha256');
+
+  files.forEach((file) => {
+    // build for every file a line looking like "aaaaaaaaaaaaaaa  file.txt\n"
+    const hash = crypto.createHash('sha256');
+
+    // a sha256sum displayed as lowercase hex string to root hash
+    const fileBuffer = fs.readFileSync(file);
+    hash.update(fileBuffer);
+    hash.end();
+    const data = hash.read();
+    rootHash.update(data.toString('hex'));
+
+    // add double space, the filename and a new line char
+    rootHash.update('  ');
+    const fileName = file.replace(/^.*[\\/]/, '');
+    rootHash.update(fileName);
+    rootHash.update('\n');
+  });
+
+  rootHash.end();
+  const rootData = rootHash.read();
+  const result: string = rootData.toString('base64');
+  return result;
+}
+
+export async function hashOfZipContent(
+  zipFilePath: string,
+  extractPath: string
+): Promise<string> {
+  await extract(zipFilePath, { dir: extractPath });
+  const files = fs.readdirSync(extractPath);
+  // the h1 hashing algorithms requires that the files are sorted by filename
+  const sortedFiles = files.sort((a, b) => a.localeCompare(b));
+  const filesWithPath = sortedFiles.map((file) => `${extractPath}/${file}`);
+
+  const result = hashFiles(filesWithPath);
+
+  // delete extracted files
+  filesWithPath.forEach((value) =>
+    fs.unlink(value, (err) => {
+      /* istanbul ignore next */
+      if (err) {
+        logger.warn({ err }, 'Failed to delete extracted file');
+      }
+    })
+  );
+  return result;
+}
+
+async function getReleaseBackendIndex(
+  backendLookUpName: string,
+  version: string
+): Promise<VersionDetailResponse> {
+  return (
+    await http.getJson<VersionDetailResponse>(
+      `${defaultRegistryUrls[1]}/${backendLookUpName}/${version}/index.json`
+    )
+  ).body;
+}
+
+export async function calculateHashes(
+  builds: TerraformBuild[],
+  cacheDir: string
+): Promise<string[]> {
+  // for each build download ZIP, extract content and generate hash for all containing files
+  const hashes = await pMap(
+    builds,
+    async (build) => {
+      const downloadFileName = `${cacheDir}/${build.filename}`;
+      const extractPath = `${cacheDir}/extract/${build.filename}`;
+      logger.trace(
+        `Downloading archive and generating hash for ${build.name}-${build.version}...`
+      );
+      const stream = http.stream(build.url);
+      const writeStream = createWriteStream(downloadFileName);
+      stream.pipe(writeStream);
+
+      const streamPromise = new Promise((resolve, reject) => {
```

Comment:
use promisify pipeline from stream here to have full async await support. add required functions to `util/fs` when missing

```js
const stream = require('stream');
const fs = require('fs');
const zlib = require('zlib');
const util = require('util');

const pipeline = util.promisify(stream.pipeline);

async function run() {
  try {
    await pipeline(
      fs.createReadStream('The.Matrix.1080p.mkv'),
      zlib.createGzip(),
      fs.createWriteStream('The.Matrix.1080p.mkv.gz'),
    );
    console.log('Pipeline succeeded');
  } catch (err) {
    console.error('Pipeline failed', err);
  }
}
```
https://nodejs.org/en/docs/guides/backpressuring-in-streams/